pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java

```diff
@@ -0,0 +1,275 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Compaction will go through the topic in two passes. The first pass
+ * selects latest offset for each key in the topic. Then the second pass
+ * writes these values to a ledger.
+ *
+ * <p>The two passes are required to avoid holding the payloads of each of
+ * the latest values in memory, as the payload can be many orders of
+ * magnitude larger than a message id.
+*/
+public class TwoPhaseCompactor extends Compactor {
+    private static final Logger log = LoggerFactory.getLogger(Compactor.class);
+    private static final int MAX_OUTSTANDING = 500;
+    private static final String COMPACTED_TOPIC_LEDGER_PROPERTY = "CompactedTopicLedger";
+
+    public TwoPhaseCompactor(ServiceConfiguration conf,
+                             PulsarClient pulsar,
+                             BookKeeper bk,
+                             ScheduledExecutorService scheduler) {
+        super(conf, pulsar, bk, scheduler);
+    }
+
+    @Override
+    protected CompletableFuture<Long> doCompaction(RawReader reader, BookKeeper bk) {
+        return phaseOne(reader).thenCompose(
+                (r) -> phaseTwo(reader, r.from, r.to, r.latestForKey, bk));
+    }
+
+    private CompletableFuture<PhaseOneResult> phaseOne(RawReader reader) {
+        Map<String,MessageId> latestForKey = new HashMap<>();
+
+        CompletableFuture<PhaseOneResult> loopPromise = new CompletableFuture<>();
+        phaseOneLoop(reader, Optional.empty(), Optional.empty(), latestForKey, loopPromise);
+        return loopPromise;
+    }
+
+    private void phaseOneLoop(RawReader reader,
+                              Optional<MessageId> firstMessageId,
+                              Optional<MessageId> lastMessageId,
+                              Map<String,MessageId> latestForKey,
+                              CompletableFuture<PhaseOneResult> loopPromise) {
+        CompletableFuture<RawMessage> future = reader.readNextAsync();
+        scheduleTimeout(future);
+        future.whenComplete(
+                (m, exception) -> {
+                    try {
+                        if (exception != null) {
+                            if (exception instanceof TimeoutException
+                                && firstMessageId.isPresent()) {
+                                loopPromise.complete(new PhaseOneResult(firstMessageId.get(),
+                                                                        lastMessageId.get(),
+                                                                        latestForKey));
+                            } else {
+                                loopPromise.completeExceptionally(exception);
+                            }
+                            return;
+                        }
+
+                        MessageId id = m.getMessageId();
+                        String key = extractKey(m);
+                        latestForKey.put(key, id);
+
+                        phaseOneLoop(reader,
```

Comment:
it might be worth checking if `loopPromise` is already completed (via cancel). If the loopPromise is already completed, you don't need to read next message.

Reply:
done